File: python/relay.py
import socket
class Relay:
    def __init__(self):
        self.port = 8080
        self.host = "192.168.1.11"
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR , 1)
        self.socket.bind((self.host, self.port))
        self.socket.listen()
        self.conn, self.addr = self.socket.accept()
    
    def write(self, msg):
        size = len(msg)
        self.conn.sendall(size.to_bytes(4, "little"))
        self.conn.sendall(str.encode(msg, encoding='ascii'))
    
    def shutdown_connection(self):
        self.conn.close()

    def shutdown_server_socket(self):
        self.socket.close()
        

File: include/Relay.cpp
#include "Relay.hpp"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <unistd.h>

#define PORT 8080

namespace hermes
{
    Relay::Relay()
    {
        int opt = 1, backlog = 5;
        printf("Beginning\n");
        this->server_fd = socket(AF_INET,SOCK_STREAM, 0);
        if(this->server_fd < 0)
        {
            printf("Socket failed\n");
            exit(EXIT_FAILURE);
        }
        // Forcefully attaching socket to the port 8080
        if (setsockopt(this->server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)))
        {
            printf("Failure to perform setsockopt\n");
            perror("setsockopt");
            exit(EXIT_FAILURE);
        }
        this->address.sin_family = AF_INET;
        this->address.sin_addr.s_addr = INADDR_ANY;
        this->address.sin_port = htons(PORT);
        printf("Address created successfully\n");

        if (bind(this->server_fd, (sockaddr *)&this->address, sizeof(this->address)) < 0)
        {
            printf("Bind operation failed\n");
            exit(EXIT_FAILURE);
        }
        printf("Bind successful\n");

        if (listen(this->server_fd, backlog) < 0)
        {
            printf("Listening operation failed\n");
            exit(EXIT_FAILURE);
        }
        printf("Listening successful\n");

        int addrlen = sizeof(this->address);
        if ((this->new_socket = accept(this->server_fd, (struct sockaddr*)&this->address,
                      (socklen_t*) &addrlen)) < 0)
        {
            perror("accept");
            exit(EXIT_FAILURE);
        }
        printf("Relay instance created successfully\n");
    }
    int Relay::Shutdown()
    {
        int close_status, shutdown_status;
        close_status = close(this->new_socket);
        shutdown_status = shutdown(this->server_fd, SHUT_RDWR);
        return close_status + shutdown_status;
    }
}
